newrelic-agent/src/main/java/com/newrelic/agent/LifecycleObserver.java

```diff
@@ -0,0 +1,142 @@
+ */
+public class LifecycleObserver {
+    protected LifecycleObserver() {
+    }
+
+    void agentStarted() {
+    }
+
+    void serviceManagerStarted(ServiceManager serviceManager) {
+    }
+
+    void agentAlreadyRunning() {
+    }
+
+    public boolean isAgentSafe() {
+        return true;
+    }
+
+    boolean isDiscovery() {
+        return false;
+    }
+
+    public static LifecycleObserver createLifecycleObserver(String agentArgs) {
+        if (agentArgs != null && !agentArgs.isEmpty()) {
+            try {
+                final AgentArguments args = AgentArguments.fromJsonObject(new JSONParser().parse(agentArgs));
+                final Number port = args.getServerPort();
+                final StatusClient client = StatusClient.create(port.intValue());
+                client.write(StatusMessage.info(args.getId(), "Msg",
+                        args.isDiscover() ? "Discovering environment" : "Initializing agent"));
+                return new AttachLifecycleObserver(client, args);
+            } catch (ParseException | IOException e) {
+                // ignore
+            }
+        }
+        return new LifecycleObserver();
+    }
+
+    private static class AttachLifecycleObserver extends LifecycleObserver {
+
+        private final StatusClient client;
+        private final AtomicReference<ServiceManager> serviceManager = new AtomicReference<>();
+        private final String id;
+        private final boolean discovery;
+
+        public AttachLifecycleObserver(StatusClient client, AgentArguments args) {
+            this.client = client;
+            this.id = args.getId();
+            this.discovery = args.isDiscover();
+        }
+
+        @Override
+        boolean isDiscovery() {
+            return discovery;
+        }
+
+        public boolean isAgentSafe() {
+            if (IBMUtils.isIbmJVM() && 
+                    !Boolean.parseBoolean(SystemPropertyFactory.getSystemPropertyProvider()
+                            .getSystemProperty(BootstrapAgent.TRY_IBM_ATTACH_SYSTEM_PROPERTY))) {
+                writeMessage(StatusMessage.error(id, "Error",
+                        "The agent attach feature is not supported for IBM JVMs"));
+                return false;
+            }
+            return true;
+        }
+
+        /**
+         * Busy waits until the agent establishes a connection with New Relic.
+         *
+         * Under normal circumstances this can take several minutes. With {@code sync_startup: true} it should be nearly instantaneous.
+         */
+        @Override
+        void agentStarted() {
+            writeMessage(StatusMessage.warn(id, "Msg",
+                    "The agent has started and is connecting to New Relic. This may take a few minutes."));
+            while (!writeConnectMessage()) {
+                try {
+                    TimeUnit.SECONDS.sleep(30);
+                    writeMessage(StatusMessage.warn(id, "Msg", "Establishing a connection with New Relic..."));
```

Comment:
This busy wait loop feels a bit gross but it works. The problem it solves is that unless the agent is configured with `sync_startup: true` it will take the agent 2-4 minutes to connect to New Relic. I think the LSI should set `sync_startup: true` by default but that's a separate issue.

Without the loop (with how the [code was structured on the prototype](https://github.com/newrelic/newrelic-java-agent/blob/c473a0f058bd8b56afaddd899e1033ed12170681/newrelic-agent/src/main/java/com/newrelic/agent/LifecycleObserver.java#L100-L105)) we would hit the code path once, immediately print out `The agent started but was not able to connect to New Relic to send data` and then nothing after even though the agent would eventually go on to successfully connect.

**Best case scenario:** With the current solution and the agent configured with `sync_startup: true` the agent immediately connects and logs the following:

```
Attaching the New Relic java agent
2021-02-17T18:03:34,484-0800 [6863 21] com.newrelic INFO: Configuration file not found. The agent will attempt to read required values from environment variables.
2021-02-17T18:03:34,505-0800 [6863 21] com.newrelic INFO: Using configured collector host: staging-collector.newrelic.com
2021-02-17T18:03:34,579-0800 [6863 21] com.newrelic INFO: New Relic Agent: Writing to log file: /private/var/folders/hp/dxbvtb094l977bx9zdhspnyc0000gn/T/logs/newrelic_agent.log
Msg	| The agent has started and is connecting to New Relic. This may take a few minutes.
Url	| https://staging.newrelic.com/accounts/123/applications/123
```

LSI output:

```
[{"agentArgs":"eJxVj8FqwzAQRP9lz25waWzLvpkgSGnimtil5GSU1ToIFClIakoJ/vdub81h5/TmsXMHbSL6GwVoZmUjZUDuZoJ3F3IJmjt08nM6yN3rZtq+DyM0EJM6G3d+Qm8tYfJh5eg7kDW4Qn+B7F+j7fupa/eSW9ufK4XWWo8PBJ/sBjm9ySND5axQi1rpWZVqXc85qmek4qQFnqpKr7GsclHgo2I4dhxjexg/enak8EWwZGD0n1CULwxHCryw94EXiSKvM7gGz/8kQ5E3LssvaFBT5w==","success":true,"messages":[{"level":"INFO","label":"Msg","message":"Initializing agent"},{"level":"WARNING","label":"Msg","message":"The agent has started and is connecting to New Relic. This may take a few minutes."},{"level":"INFO","label":"Url","message":"https://staging.newrelic.com/accounts/123/applications/123"}],"pid":6863,"command":null}]
```

**Worst case scenario:** With the current solution and the agent configured with `sync_startup: false` the agent takes several minutes to connect, running through the busy loop a few times (each time taking 30 seconds before it checks the connection and prints again), and logs the following:

```
Attaching the New Relic java agent
2021-02-17T18:13:12,797-0800 [7044 21] com.newrelic INFO: Configuration file not found. The agent will attempt to read required values from environment variables.
2021-02-17T18:13:12,815-0800 [7044 21] com.newrelic INFO: Using configured collector host: staging-collector.newrelic.com
2021-02-17T18:13:12,881-0800 [7044 21] com.newrelic INFO: New Relic Agent: Writing to log file: /private/var/folders/hp/dxbvtb094l977bx9zdhspnyc0000gn/T/logs/newrelic_agent.log
Msg	| The agent has started and is connecting to New Relic. This may take a few minutes.
Msg	| Establishing a connection with New Relic...
Msg	| Establishing a connection with New Relic...
Msg	| Establishing a connection with New Relic...
Msg	| Establishing a connection with New Relic...
Msg	| Establishing a connection with New Relic...
Url	| https://staging.newrelic.com/accounts/123/applications/123
```

LSI output:

```
[{"agentArgs":"eJxVj8sKwjAQRf9l1lUq9L0rElDUWlQQVyVOphJIk5IURaT/7ix1cXf33McHlA7onuSh6qUJFAHZp/bODmQnqD7QiGt3EvvtutsczxeoIEzyoe1jgc4Ywsn5paWXJ6NxiW6A6Ieo27Zr6oNgavMeydfGOPxzsERzFt1O3NiU9RJVUUrVy0wmZR+jXCGld1XgPc9VglkeFykizBFoxUAeJwnnBfL8oHWeFxdpXEYwesd9k6bAH+b5C53DSks=","success":true,"messages":[{"level":"INFO","label":"Msg","message":"Initializing agent"},{"level":"WARNING","label":"Msg","message":"The agent has started and is connecting to New Relic. This may take a few minutes."},{"level":"WARNING","label":"Msg","message":"Establishing a connection with New Relic..."},{"level":"WARNING","label":"Msg","message":"Establishing a connection with New Relic..."},{"level":"WARNING","label":"Msg","message":"Establishing a connection with New Relic..."},{"level":"WARNING","label":"Msg","message":"Establishing a connection with New Relic..."},{"level":"WARNING","label":"Msg","message":"Establishing a connection with New Relic..."},{"level":"INFO","label":"Url","message":"https://staging.newrelic.com/accounts/123/applications/123"}],"pid":7044,"command":null}]
```